pep-9999.rst

```diff
@@ -0,0 +1,257 @@
+PEP: 9999
+Title: Package index configuration standardization
+Author: Fredrik Averpil <fredrik at averpil.com>
+Status: Draft
+Type: Informational
+Created: 05-May-2021
+Post-History: <dates of postings to python-ideas and/or python-dev>
+
+
+Abstract
+========
+
+This PEP standardizes configuration of PyPI repositories.
+
+
+Motivation
+==========
+
+Workflow tools come up with their own configuration mechanisms for PyPI repositories, as there is no standard for this. This causes fragmentation and pain points.
+
+Examples of such configuration files:
+
+* Pip's ``pip.conf`` and ``pip.ini``
+* Packaging's ``.pypirc`` (also used by tools like Flit and Twine)
+* Poetry's ``config.toml`` and ``auth.toml``
+
+
+Rationale
+=========
+
+While this PEP aim to come up with a unified PyPI repository configuration, thought has been given so to be able to extend the configuration later. For example, PyPI repositories are assumed in this PEP, but a ``type`` attribute could be introduced which would default to ``pypi`` down the line, to allow for other package repository types. Any tool could also start taking advantage of the fact that the configuration files and their discovery is a standard, and use this for their own configuration.
+
+
+Specification
+=============
+
+The specification consists of three parts: the configuration definition, package discovery and configuration file discovery.
+
+
+Configuration definition
+------------------------
+
+The configuration is split into two files:
+
+* ``config.toml`` for PyPI repository configuration
+* ``auth.toml`` for credentials
+
+
+The config.toml file
+''''''''''''''''''''
+
+Package index configuration is defined in the ``config.toml`` file::
```

Comment:
Should these be also supported as ENV VARS?

Reply:
Yeah, I've been thinking of that many times. I mean, why not. All it takes is to settle on the naming of such an environment variable 😉 

The env var should then point to a folder which should contain the `config.toml` and/or `auth.toml` files, I guess.

Reply:
[Here](https://docs.python.org/3/using/cmdline.html#environment-variables) are all the env vars specified in the official docs.

Seems dangerous to claim e.g. `PYTHONCONFIG`. 😉

Reply:
Maybe claim a prefix? `PYTHONINDEX` or something? Depends on if we defining this for index config alone. If we are we should probably also not recommend `config.toml` as the name :)

Reply:
So far I think we are open to have other tools make use of this `config.toml` file, in other areas than package index configuration. Because of this, I don't think it's good to limit ourselves here with the choice of the environment variable name 😅 

I assume whatever environment variable name we choose here, it won't end up being added to the official Python documentation (?). Instead I'd assume it would be limited to be specified here in this PEP. And therefore, I don't think we should use a variable name which signals it _could be_ an officially Python-supported env var.

Having all this said, I think it could be good to offer this behavior. I can see how this could be useful in CI situations. And I'd say it should be evaluated last, after the per-project settings. This way, it could override everything stipulated in any of the other file locations and act as the "final verdict" on e.g. whether an index is enabled or not.

Reply:
@domdfcoding do you have any input on naming for an env var which would point to a path which would hold the `config.toml` and/or `auth.toml`?

I was also thinking this env var would cause configuration detection to be skipped. Meaning, no files which may be placed in the differrent operating system locations specified in this PEP (global, per-user, per-environment) would be read.

Makes sense if this would be something like `$PYTHON_CONFIG_PATH`.

Optionally, another env var could instruct the config detection to use this user-defined `$PYTHON_CONFIG_PATH` at the end of the regular config discovery, so that any config in this path would be added on top of the regular config discovery. For example, we could enable this behavior if the user also specifies `$PYTHON_CONFIG_PATH_ACC` (acc for accumulation).